Meter.c

```diff
@@ -144,11 +148,97 @@ ListItem* Meter_toListItem(Meter* this, bool moving) {
    return li;
 }
 
+/* ---------- GraphData ---------- */
+
+static GraphData* GraphData_new(size_t nValues, size_t nItems, bool isPercentGraph) {
+   // colors[] is designed to be two-dimensional, but without a column of row
+   // pointers (unlike var[m][n] declaration).
+   // GraphMeterMode_draw() will print this table in 90-deg counter-clockwise
+   // rotated form.
+   unsigned int colorRowSize;
+   if (nItems <= 1) { // 1 or less item
+      colorRowSize = 1;
+   } else if (isPercentGraph) { // Percent graph: a linear row of color cells.
+      colorRowSize = GRAPH_HEIGHT;
+   } else { // Non-percentage & dynamic scale: a binary tree of cells.
+      colorRowSize = 2 * GRAPH_HEIGHT;
+   }
+   GraphData* data = xCalloc(1, sizeof(GraphData) +
+                                sizeof(double) * nValues +
+                                sizeof(double) * (nItems + 2) * 2 +
+                                sizeof(int) * (nValues * colorRowSize));
```

Comment:
Missing overflow check in calculation of allocated size. Also mixing allocations of different sizes into one makes maintaining the source unnecessarily difficult.

Reply:
@BenBE 
If the data to be allocated is from user input, I would have already considered the size overflow problem seriously. But for this case, how come would a raster  graph buffer take even a megabyte of memory space? Sorry, but I don't believe there would be a security issue here.

Second, there is no alignment problem here. C compiler handles struct alignment automatically, and misalignment can only happen if you override the alignment (via `#pragma pack` and such) or you cast a pointer in order to do an unaligned read.

Reply:
As mentioned, I'm mostly over-cautious in this regard and regard combining allocations as bad style if no clear necessity for doing so can be given.

Regarding the alignment: Just after the allocation comes code, that casts the memory buffers around. The types are fine (and alignment only gets smaller), but it's a pitfall waiting for someone to trip over.